prometheus/src/cluster_metrics.rs

```diff
@@ -1,12 +1,84 @@
 use solana_gossip::cluster_info::ClusterInfo;
 use solana_runtime::bank::Bank;
+use solana_sdk::{clock::Slot, pubkey::Pubkey};
+use solana_vote_program::vote_state::VoteState;
 
 use crate::{
     token::Lamports,
     utils::{write_metric, Metric, MetricFamily},
 };
 use std::{io, sync::Arc, time::SystemTime};
 
+struct ValidatorVoteInfo {
+    vote_address: Pubkey,
+    balance: Lamports,
+    last_vote: Slot,
+}
+
+impl ValidatorVoteInfo {
+    fn new_from_bank(bank: &Arc<Bank>, identity_pubkey: &Pubkey) -> Option<Self> {
+        let vote_accounts = bank.vote_accounts();
+        let vote_state_default = VoteState::default();
+        vote_accounts
+            .iter()
+            .filter_map(|(&vote_pubkey, (_activated_stake, account))| {
+                let vote_state = account.vote_state();
+                let vote_state = vote_state.as_ref().unwrap_or(&vote_state_default);
+                if identity_pubkey != &vote_state.node_pubkey {
+                    return None;
+                }
+                let last_vote = if let Some(vote) = vote_state.votes.iter().last() {
+                    vote.slot
+                } else {
+                    0
+                };
+                let vote_balance = Lamports(bank.get_balance(&vote_pubkey));
+                Some(ValidatorVoteInfo {
+                    vote_address: vote_pubkey,
+                    balance: vote_balance,
+                    last_vote,
+                })
+            })
+            .next()
+    }
+
+    fn write_prometheus<W: io::Write>(&self, out: &mut W, at: SystemTime) -> io::Result<()> {
```

Comment:
Like in the previous PR, we sample everything just in time as Prometheus requests it, so we don’t need this `at`.